python/lsst/afw/geom/endpoint.cc

```diff
@@ -0,0 +1,165 @@
+// This is meant to be called by other `declare...` functions;
+template <typename Point>
+void declareBaseVectorEndpoint(py::module& mod, std::string const& suffix) {
+    using Class = BaseVectorEndpoint<Point>;
+    using Array = typename Class::Array;
+
+    declareBaseEndpoint<Point, Array>(mod, suffix);
+
+    py::class_<Class, std::shared_ptr<Class>, BaseEndpoint<Point, Array>> cls(
+            mod, ("_BaseVectorEndpoint" + suffix).c_str());
+
+    cls.def(py::init<int>(), "nAxes"_a);
+
+    cls.def("getNPoints", &Class::getNPoints);
+    cls.def("dataFromPoint", &Class::dataFromPoint);
+    cls.def("dataFromArray", &Class::dataFromArray);
+    cls.def("arrayFromData", &Class::arrayFromData);
+    cls.def("pointFromData", &Class::pointFromData);
+}
+
+// Declare GenericEndpoint and all subclasses
+void declareGenericEndpoint(py::module& mod) {
+    using Class = GenericEndpoint;
+    using Point = typename Class::Point;
+    using Array = typename Class::Array;
+
+    declareBaseEndpoint<Point, Array>(mod, "Generic");
+
+    py::class_<Class, std::shared_ptr<Class>, BaseEndpoint<Point, Array>> cls(mod, "GenericEndpoint");
+
+    cls.def(py::init<int>(), "nAxes"_a);
+
+    cls.def("getNPoints", &Class::getNPoints);
+    cls.def("dataFromPoint", &Class::dataFromPoint);
+    cls.def("dataFromArray", &Class::dataFromArray);
+    cls.def("arrayFromData", &Class::arrayFromData);
+    cls.def("pointFromData", &Class::pointFromData);
+}
+
+/// declare PointNEndpoint (for N = 2 or 3) and all subclasses
+template <int N>
+void declarePointEndpoint(py::module& mod) {
+    using Class = PointEndpoint<N>;
+    using Point = typename Class::Point;
+    std::stringstream os;
+    os << "Point" << N;
+    std::string pointNumStr = os.str();
+
+    declareBaseVectorEndpoint<Point>(mod, pointNumStr);
+
+    py::class_<Class, std::shared_ptr<Class>, BaseVectorEndpoint<Point>> cls(
+            mod, (pointNumStr + "Endpoint").c_str());
+
+    cls.def(py::init<>());
+    // do not wrap the constructor that takes nAxes; it is an implementation detail
+
+    cls.def("normalizeFrame", &Class::normalizeFrame);
+}
+
+/// declare SpherePointEndpoint and all subclasses
+void declareSpherePointEndpoint(py::module& mod) {
+    using Class = SpherePointEndpoint;
+    using Point = typename Class::Point;
+
+    declareBaseVectorEndpoint<Point>(mod, "SpherePoint");
+
+    py::class_<Class, std::shared_ptr<Class>, BaseVectorEndpoint<Point>> cls(mod, "SpherePointEndpoint");
+
+    cls.def(py::init<>());
+    // do not wrap the constructor that takes nAxes; it is an implementation detail
+
+    cls.def("normalizeFrame", &Class::normalizeFrame);
+    // return a deep copy so Python cannot modify the internal state
+    cls.def("makeFrame", [](Class const & self) {
+        auto frame = self.makeFrame();
+        return frame->copy();
+    });
+}
+}  // <anonymous>
+
+PYBIND11_PLUGIN(_endpoint) {
+    // TODO uncomment these imports during the cleanup pass
```

Comment:
It is waiting for the wrapper library name to not have the leading underscore, which is part of cleaning up the afw.geom pybind11 wrapper. I'll clarify the comment.